### ISSUE TITLE
[MU4 Issue] Palette search excludes elements behind "More" button

### ISSUE THREAD

**Issue by shoogle** [musescore/MuseScore#11113]
**Describe the bug**

To avoid cluttering the palettes, some elements are hidden behind a "More" button. Currently these elements do not appear in search results, yet elements from hidden palettes (the ones behind the "Add palettes" button) do appear in search results.

For the sake of discoverability, searches in the palettes should return all matching elements even if those elements are not usually visible in the palettes.

**Example**

The "More" list of the Articulations palette contains an element called "Snap pizzicato above", but this item does not appear in search results. We want it to appear.

"Snap pizzicato above" exists in Articulations > More | Searching for "snap" does not reveal it
---|---
<img width="428" alt="image" src="https://user-images.githubusercontent.com/11011881/162117534-046a8268-daec-4176-9008-fcd5d5d9e3aa.png"> | <img width="301" alt="image" src="https://user-images.githubusercontent.com/11011881/162118117-23770bbd-040f-4991-a2f6-7f3aa9e680d8.png">

### Other considerations

When fixing the above, we want to **preserve the following behaviour** which is correct as it is now.

**Add palettes**

Searching for "grace" returns elements from the Bagpipe embellishments palette even though this palette is hidden behind the "Add palettes" button. We want this behaviour to continue.

"Bagpipe embellishments" exists in the 'Add palettes' popup | Searching for "grace" includes Bagpipe embellishments
---|---
<img width="301" alt="image" src="https://user-images.githubusercontent.com/11011881/162120012-5ad8a781-21c4-484a-8268-5dcff47b68fd.png"> | <img width="301" alt="image" src="https://user-images.githubusercontent.com/11011881/162118866-f71b4fbc-4d1f-41a9-825d-86cf8a545328.png">

**Custom palettes**

Users can create or load custom palettes and drag additional elements into those palettes. These elements are included in search results even though they may also appear in other palettes. We want this to continue.

Here "Trill" has been added to a custom palette | Searching for "trill" includes the custom palette
---|---
<img width="301" alt="image" src="https://user-images.githubusercontent.com/11011881/162588367-dc7083e0-9ad7-4850-b256-b17da2bdae5a.png"> | <img width="301" alt="image" src="https://user-images.githubusercontent.com/11011881/162588496-e76b7052-65ab-4d13-b331-6c09aa4d1313.png">

Basically, we want palette searches to include everything!

**Comment by mccoydtromb** [comment]
I have a potential fix by using the master palette model to search instead of the user palette model, but it doesn't hide results from palettes that have not been added yet. Is that acceptable?

**Comment by MarcSabatella** [comment]
It’s useful, but not really the same - the palattes can also contain elements added by the user, and search should encompass those too.

**Comment by mccoydtromb** [comment]
Ah, didn't know that. I'll see what I can do. Sounds like we might need a search model that can encompass all of these, or else be able to choose between searching the master palette and the user palette.

**Comment by shoogle** [comment]
@mccoydtromb, thanks for the PR!

> it doesn't hide results from palettes that have not been added yet. Is that acceptable?

Yes, that's actually what we want. I have updated the issue description to make this more clear.

**Comment by shoogle** [comment]
> palettes can also contain elements added by the user, and search should encompass those too.

That's right. I've updated the issue description to mention custom palettes.

> Sounds like we might need a search model that can encompass all of these, or else be able to choose between searching the master palette and the user palette.

One option might be to put all elements in the User palette and then just set a 'hidden' flag (or remove a 'visible' flag) on any that are supposed to go behind a 'More' button.

**Comment by Eism** [comment]
I researched this problem and discussed with Roman.
It looks like the fix(not a workaround) will affect a lot of the code in the pallets. This is very dangerous to do before release.

Also I couldn't find a workaround(which does not break pallets) to fix this bug.

We need to choose:
1. Move to 4.x. I know that there is a problem in MU3 too and for MU4 it is more critical, but at the moment we can only find a workaround, rather than fix the problem correctly.
2. Merge #11126. Then the user will not find custom pallets in the search. Is it critical than the main issue? 

@Tantacrul FYI

**Comment by MarcSabatella** [comment]
How about 3) provide a "full" workspace option that simply includes all elements on the palettes directly, nothing hidden behind "more".  Then a user who relies primarily on search can opt to select that worksapce.

The idea of providing additional workspaces to select between is one I strongly support anyhow - although if not for this issue, I wouldn't have specifically thought this "full" option should be one of them.  See also https://github.com/musescore/MuseScore/issues/11057

**Comment by shoogle** [comment]
@MarcSabatella, I support having a "show me everything" workspace but I don't see that as a way to solve search (which should return all relevant results regardless of which workspace you are in).

I suppose the act of searching could temporarily switch you to the "everything" workspace, but that wouldn't help with custom palettes so it is essentially equivalent to option (2).

**Comment by MarcSabatella** [comment]
Well, the division of what's shown initially versus what is behind More is only defined relative to the current workspace.  Same with custom palettes.  If we're saying palette search currently searches custom palettes (in the current workspace) but ignores things behind More (in the current workspace), then using the proposed Full workspace would have to work; I can't see how it could not.  Not to temporarily switch to it upon doing the search - I mean, a user who wanted to rely primarily on palette search would switch to the Full workspace immediately upon installing MuseScore.  If they wanted to customize, they'd, do all their customizations there as well.

Users can of course solve the problem for themselves already by taking the same approach - laboriously creating their own Full workspace by copying everything under More to the main palette.  This works perfectly (other than the fact that palette search crashes for me more often than not).

**Comment by shoogle** [comment]
@MarcSabatella, correct. Users who opt-in to a Full workspace would not face this issue. However, the issue remains for all users who do not opt-in to the Full workspace, so the Full workspace is not a solution to this issue.

**Comment by Tantacrul** [comment]
I agree with @shoogle that option 2 makes a **lot** more sense. It is far less likely to be experienced by users and will cover way more elements. So can we do the following @Eism?

> Merge https://github.com/musescore/MuseScore/pull/11126. Then the user will not find custom pallets in the search. Is it critical than the main issue?

**Comment by mccoydtromb** [comment]
For 4.x, I agree with Marc that it should be possible to create a search filter model that is a union of master palette and user palette, and that is what I intended to do, but my C++ skills are too rusty in this area. I'll make the change in the current nightly and resubmit the PR, but I'll happily accept guidance from a coding master on creating a new search filter model for 4.x

**Comment by shoogle** [comment]
@mccoydtromb, feel free to work on a complete fix for 4.x but please do that on a separate PR. Your current PR + minor adjustments is all that we are willing to consider for 4.0.
